### DISCUSSION TITLE
How to force notifying DataContext changed?

### DISCUSSION THREAD

**Discussion by liwuqingxin** [AvaloniaUI/Avalonia#15229] [Q&A]
Hello everyone!

EDIT 1 --------

More broadly, the question is how to force to notify an AvaloniaProperty changed, even it is not changed actually.
For an instance that implemented INotifyPropertyChanged, we can use instance.PropertyChanged(...) to do that. But with AvaloniaObject's AvaloniaProperty, how to do that?

END EDIT 1 --------

I have a custom DataGrid control. I use static expression like 'value / sum(value)' to calculate some column's value. In that situation, the calculation only performs when the DataContext of the row, which is also the cells'DataContext, is changed or the control is creating.

Now when I update a rowdata.value from '10' to '20', I want to force to notify that the DataRow's DataContext is changed, to update the expression column's cell. How can I do that efficiently?

I try to set the DataContext of the DataRow to null first and then set it back immediately. It works, but is not efficient enough. The DataContext updated twice in that way.

So, I would like to find out an api that is like INotifyPropertyChanged.PropertyChanged, which can notify the AvaloniaProperty change even if it is not changed actually.

Many thanks!

![image](https://github.com/AvaloniaUI/Avalonia/assets/20198528/a7dfef92-96e3-4c54-be03-da93fe5c200a)

**Comment by timunie** [comment]
that's a task for your ItemViewModel. `this.RaisePropertyChanged(nameof(TheOtherProperty))` if you use ReactiveUI. If you use another MVVM toolkit, lookup their docs.

**Reply by liwuqingxin** [reply]
No, any property of the Row Item changed can not notify the cell to update it's value calculated by expression. The expression cell's value dependent on the DataContext but it's properties. Only the DataGridRow's DataContext changed from an object to another object, the expression cell's value changed.

**Reply by liwuqingxin** [reply]
The summary cell control is created like this.  So, **The expression cell's value dependent on the DataContext property but it's properties.**

```csharp
    protected virtual Control? OnCreateSummaryControl()
    {
        if (SummaryDataTemplate != null)
        {
            ......
        }
        else
        {
            // TODO 这里待修改
            var control = new LightTextBlock();
        
            control.FontWeight = FontWeight.Bold;
            control.FontSize   = 14;
        
            var binding = new Binding
            {
                //
                // No Source set here. Use DataContext as source.
                //
                Path     = nameof(TableRowGroupModel.Context),
                Priority = BindingPriority.Style,
                Converter = new DynamicCvt((value, type, parameter, culture) =>
                {
                    if (value is TableRowGroupContext context)
                    {
                        var summaryContext = new ExpressionContext<double>();
                        summaryContext.Item  = context;
                        summaryContext.Items = context.OriginalRows?.Select(r => r.Item).ToList();
                        var ret = CompiledSummaryExpression?.Calc(summaryContext);
                        return $"{ret:N0}";    
                        return $"{ret:f2}";    
                    }

                    return BindingOperations.DoNothing;
                }),
            };
            control.Bind(TextBlock.TextProperty, binding);
        
            return control;
        }
    }
```

**Reply by liwuqingxin** [reply]
All I need is to notify some control's DataContext is changed, even it is not changed actrually. 
Setting the control's DataGrid = null and then setting it back performs well, but not efficiently.

**Reply by timunie** [reply]
`this.RaisePropertyChanged(string.Empty)` --> That should trigger the Binding for all properties

**Reply by liwuqingxin** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/20198528/5657fe5f-77a6-444f-a350-5975cb84ed5a)

DataContextProperty is AvaloniaProperty but DirectProperty. RaisePropertyChanged does not fit for that. And I have tried to call the overload method of RaisePropertyChanged by reflecting. But it does not take effect.

```csharp
 internal void RaisePropertyChanged<T>(
            AvaloniaProperty<T> property,
            Optional<T> oldValue,
            BindingValue<T> newValue,
            BindingPriority priority,
            bool isEffectiveValue);
```

**Reply by timunie** [reply]
On your ViewModel, leave the control alone. It's not how MVVM works

**Reply by liwuqingxin** [reply]
I have no ViewModel... I am creating a TableControl like DataGrid In Avalonia. So in my custom table control, I don't care about the ViewModel that the user provided. I Just wants to get notifed when any property of the row data changed and then notify the expression column to perform it's calculation **again**.

![image](https://github.com/AvaloniaUI/Avalonia/assets/20198528/c739520a-c142-4d9c-b3f1-87d20bbd7493)

**Reply by liwuqingxin** [reply]
The StaticExpressionColumn binds to the DataContext and use a converter to calculate expression. So only  the DataContext is changed the expression can be calculated again.

```csharp
public class StaticExpressionColumn : DoubleBasedExpressionColumn
{
    protected override Control OnCreateCellControl()
    {
        var control = new LightTextBlock();
        var binding = new Binding
        {
            Priority = BindingPriority.LocalValue,
            Converter = new DynamicCvt((value, type, parameter, culture) =>
            {
                var ret = value == null ? value : CalcRawValue(value);
                return $"{ret:f2}";
            }),
        };
        control.Bind(TextBlock.TextProperty, binding);
        
        return control;
    }

    protected override string? GetFallbackHeader()
    {
        return Expression;
    }
}
```
